node-syscall/syscall.cc

```diff
@@ -23,76 +56,96 @@ intptr_t toNative(Local<Value> value) {
   }
   if (value->IsArray()) {
     Local<Array> array = Local<Array>::Cast(value);
-    intptr_t* native = reinterpret_cast<intptr_t*>(malloc(array->Length() * sizeof(intptr_t))); // TODO memory leak
+    intptr_t* native = a.allocate(array->Length());
     for (uint32_t i = 0; i < array->Length(); i++) {
-      native[i] = toNative(array->Get(i));
+      native[i] = toNative(ctx, a, array->Get(ctx, i).ToLocalChecked());
     }
     return reinterpret_cast<intptr_t>(native);
   }
-  return static_cast<intptr_t>(static_cast<int32_t>(value->ToInteger()->Value()));
+
+  return static_cast<intptr_t>(static_cast<int32_t>(integerOrDie(ctx, value)->Value()));
 }
 
 void Syscall(const FunctionCallbackInfo<Value>& info) {
-  int trap = info[0]->ToInteger()->Value();
-  int r1 = 0, r2 = 0;
-  switch (trap) {
-  case SYS_fork:
-    r1 = fork();
-    break;
-  case SYS_pipe:
-    int fd[2];
-    r1 = pipe(fd);
-    if (r1 == 0) {
-      r1 = fd[0];
-      r2 = fd[1];
+  arena a;
+  Isolate* isolate = info.GetIsolate();
+  Local<Context> ctx = isolate->GetCurrentContext();
+
+  try {
+    int trap = integerOrDie(ctx, info[0])->Value();
+    int r1 = 0, r2 = 0;
+    switch (trap) {
+    case SYS_fork:
+      r1 = fork();
+      break;
+    case SYS_pipe:
+      int fd[2];
+      r1 = pipe(fd);
+      if (r1 == 0) {
+        r1 = fd[0];
+        r2 = fd[1];
+      }
+      break;
+    default:
+      r1 = syscall(
+        trap,
+        toNative(ctx, a, info[1]),
+        toNative(ctx, a, info[2]),
+        toNative(ctx, a, info[3])
+      );
+      break;
     }
-    break;
-  default:
-    r1 = syscall(
-      trap,
-      toNative(info[1]),
-      toNative(info[2]),
-      toNative(info[3])
-    );
-    break;
-  }
-  int err = 0;
-  if (r1 < 0) {
-    err = errno;
+    int err = 0;
+    if (r1 < 0) {
+      err = errno;
+    }
+    Local<Array> res = Array::New(isolate, 3);
+    // TODO: Actually check the value.
```

Comment:
Does this TODO need to be resolved before this PR is merged, or can that be done later?

Also, I don't quite understand what the TODO means. Check the value for what? If you're planning on keeping it, please expand it so it's clear what needs to be done here.